### PR TITLE
Added full support for screenshots with custom area selection.

### PR DESCRIPTION
**Pull request by DustVoice**
To take a screenshot and insert it at the current cursor position, select a node, where you can also insert an image and then either
- Go to "Edit" then to "Insert Screenshot"
- Press the icon beside the "Insert Image" Icon
- Right click in the Editor -> "Insert" -> "Insert Screenshot"
- Press ctrl + shift + alt + s

After that a fullscreen medium gray window will appear. This darker area is where the screen gets captured for the screenshot. In the middle the current active mode is displayed. Under the mode indicator a small text tells you to double-click on it to get some help. If you do, a popup will appear that allows you to read the same instructions as listed below.

If you don't want to capture the whole screen, you now have to navigate/resize the overlay to the area you want to capture
1. With _Space_ or _Tab_ you can toggle between the "Move-Mode" and the "Resize-Mode"
- In "Move-Mode" you can drag and drop the overly by holding the left mouse button and dragging the mouse. The overlay will follow
- In "Resize-Mode" you can resize the window with the bottom-right corner, which means holding the left mouse button and dragging the mouse will move the bottom-right corner
2. If you want to make fine adjustments, you might find the other four modes to come in handy - The Edge Modes!
- If you press the _up_ arrow-key, or alternatively _w_ on your keyboard, the overlay will switch to "Top-Edge-Mode". This means that holding the left mouse button and dragging the mouse will move the upper corner of the overlay up and down
- If you press the _down_ arrow-key, or alternatively _s_ on your keyboard, the overlay will switch to "Bottom-Edge-Mode". This means that holding the left mouse button and dragging the mouse will move the lower corner of the overlay up and down
- If you press the _left_ arrow-key, or alternatively _a_ on your keyboard, the overlay will switch to "Left-Edge-Mode". This means that holding the left mouse button and dragging the mouse will move the left corner of the overlay left and right
- If you press the _right_ arrow-key, or alternatively _d_ on your keyboard, the overlay will switch to "Right-Edge-Mode". This means that holding the left mouse button and dragging the mouse will move the right corner of the overlay left and right
3. If you want to return to "Move-Mode" again, simply follow instruction number 1 and press _Space_ or _Tab_. If you want to switch to the "Resize-Mode" again, press _Space_ or _Tab_ twice.
4. If the overlay covers the exact area you want to have a screenshot of, simply press "Return" or "Enter" to take the screenshot. After that the "Edit-Image" dialog will appear, exactly as if you would have just picked an image from the image browser. resize the image if you want and press ok to insert the image
5. If you want to cancel the screenshot simply press "Esc"

### CODE REVIEW

**Comment by DustVoice** [comment]
It would be at least nice, if someone could build this for me, as I can't seem to get the build process going!

File: modules/cons.py [hunk 1]
```diff
@@ -245,7 +245,7 @@
 'node_bullet', 'node_no_icon', CHERRY_BLACK, CHERRY_BLUE,
 CHERRY_CYAN, CHERRY_GREEN, CHERRY_GRAY, CHERRY_ORANGE,
 CHERRY_ORANGE_DARK, CHERRY_PURPLE, CHERRY_RED, CHERRY_SHERBERT,
-CHERRY_YELLOW, 'image_insert', 'image_edit', 'image_save',
+CHERRY_YELLOW, 'image_insert', 'screenshot_insert', 'image_edit', 'image_save',
```
Comment (DustVoice) [inline modules/cons.py]: Add custom icon for the screenshot button

File: modules/core.py [hunk 2]
```diff
@@ -4046,12 +4051,26 @@ def image_handle(self, *args):
...
         if not filename: return
         self.pick_dir_img = os.path.dirname(filename)
+        
         try:
             pixbuf = gtk.gdk.pixbuf_new_from_file(filename)
             self.image_edit_dialog(pixbuf, self.curr_buffer.get_iter_at_mark(self.curr_buffer.get_insert()))
         except:
             support.dialog_error(_("Image Format Not Recognized"), self.window)
 
+    def screenshot_handle(self, *args):
```
Comment (DustVoice) [inline modules/core.py]: Custom handle for button click (I used image_handle as a template)

File: modules/menus.py [hunk 3]
```diff
@@ -64,6 +64,7 @@
 "act_undo",
 "act_redo",
 "handle_image",
+"handle_screenshot",
```
Comment (DustVoice) [inline modules/menus.py]: Register the new menu entry

File: modules/menus.py [hunk 4]
```diff
@@ -185,6 +186,7 @@ def load_menudict(dad):
 "act_undo": {"sk": "gtk-undo", "sd": _("_Undo"), "kb": KB_CONTROL+"Z", "dn": _("Undo Last Operation"), "cb": dad.requested_step_back},
 "act_redo": {"sk": "gtk-redo", "sd": _("_Redo"), "kb": KB_CONTROL+"Y", "dn": _("Redo Previously Discarded Operation"), "cb": dad.requested_step_ahead},
 "handle_image": {"sk": "image_insert", "sd": _("Insert I_mage"), "kb": KB_CONTROL+KB_ALT+"I", "dn": _("Insert an Image"), "cb": dad.image_handle},
+"handle_screenshot": {"sk": "screenshot_insert", "sd": _("Insert _Screenshot"), "kb": KB_CONTROL+KB_SHIFT+KB_ALT+"S", "dn": _("Insert a Screenshot"), "cb": dad.screenshot_handle},
```
Comment (DustVoice) [inline modules/menus.py]: Register the new menu entry

File: modules/menus.py [hunk 5]
```diff
@@ -487,6 +489,7 @@ def get_popup_menu_entries_text(dad):
 (cons.TAG_SEPARATOR, None, None, None, None),
 ("submenu-start", _("_Insert") , "insert", None, None),
 get_menu_item_tuple(dad, "handle_image"),
+get_menu_item_tuple(dad, "handle_screenshot"),
```
Comment (DustVoice) [inline modules/menus.py]: Register the new menu entry

File: modules/menus.py [hunk 6]
```diff
@@ -683,6 +686,7 @@ def get_popup_menu_entries_codebox(dad):
       <menuitem action='act_redo'/>
       <separator/>
       <menuitem action='handle_image'/>
+      <menuitem action='handle_screenshot'/>
```
Comment (DustVoice) [inline modules/menus.py]: Register the new menu entry

File: modules/screenshot.py [hunk 7]
```diff
@@ -0,0 +1,150 @@
+# -*- coding: UTF-8 -*-
```
Comment (DustVoice) [inline modules/screenshot.py]: Custom class for handling the screenshot

File: modules/screenshot.py [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+
+        self.label.show()
+        
+        self.add(self.label)
+        
+        self.update_mode_label(self.label)
+
+        self.fullscreen()
+
+    def run(self):
```
Comment (DustVoice) [inline modules/screenshot.py]: Dialog-like function that blocks all activity (but does not freeze the GUI), until it is finished or aborted

File: modules/support.py [hunk 9]
```diff
@@ -792,6 +793,10 @@ def dialog_folder_select(curr_folder=None, parent=None):
         chooser.destroy()
         return None
 
+def dialog_screenshot_take(parent=None):
```
Comment (DustVoice) [inline modules/support.py]: To follow the coding style, I outsourced this function to this file.

**Comment by giuspen** [comment]
Thanks for your work I will analyse soon and likely integrate unless I see problems. Bear with me I'm a really busy father of two these days struggling to find minutes of free time :)

**Comment by DustVoice** [comment]
Yeah no problem! Many happy moments with your children, I find you very inspiring. Keep up the great work and don't rush anything.

File: modules/core.py [hunk 10]
```diff
@@ -49,13 +49,18 @@
 import findreplace
 import codeboxes
 import ctdb
+import screenshot
+import gtk.gdk
+from threading import Event, Thread
```
Comment (giuspen) [inline modules/core.py]: This import doesn't seem to be used, am I right?

**Comment by DustVoice** [comment]
Oh yeah sorry about that. Gtk is already imported, so gtk.gdk shouldn't be needed and I used a different approach instead of threading to make it more stable. These are residues from testing and developing. Import screenshot should still remain. But the other two you can safely remove. Thanks for the note. Completely forgot about that

**Comment by DustVoice** [comment]
I removed the unused imports and commited the changes. Please contact me again for any further questions.

**Comment by giuspen** [comment]
If I minimise the cherrytree main window to taskbar to take a screenshot of something else without cherrytree in the way, it always crashes when I press Enter to take the screenshot (Linux Mint 18.3 Cinnamon).

**Comment by DustVoice** [comment]
Ok I will look into that. It didn't do that for win 10 or Ubuntu, but I will test it in Linux mint

**Comment by giuspen** [comment]
Thanks David, if you will have the possibility to debug it on linux that would be great, otherwise no worries I will look into that when I have a moment

**Comment by giuspen** [comment]
I fixed the crash but still struggling to have the gray overlay out of the screenshot

**Comment by DustVoice** [comment]
The gray overlay? This should never be in a Screenshot because the window is killed before taking the Screenshot... What machine are you on again?

**Comment by DustVoice** [comment]
Could you provide me with a sample Screenshot?

**Comment by giuspen** [comment]
Linux Mint Cinnamon 16.04, later I will prepare a screen shot, now family time. I also tried to destroy the whole dialog before taking the screen shot (current master version) but still got the issue.

**Comment by DustVoice** [comment]
ctree
ctree2
ctree3

As you can see the issue is not on my system. hmm. i'll have to look into it when i have time

**Comment by hexterisk** [comment]
Using the latest version but can't see it anywhere. Was this feature removed?

**Comment by giuspen** [comment]
It's not working and I have no time to debug it, it is in the code just its menu item is commented out

**Comment by hexterisk** [comment]
So the shortcut should trigger the feature? Because that's not happening either.
Also, great work with the app. Cheers!

**Comment by giuspen** [comment]
If you make it work perfectly I can include it, the efforts are currently to be bug free and port to Gtkmm3 from discontinued pygtk2

**Pull request merged**